ssl/quic/quic_impl.c

```diff
@@ -9,179 +9,2104 @@
+ *
+ */
+
+static int blocking_mode(QUIC_CONNECTION *qc)
 {
-    return -1;
+    return 1; // TODO
+}
+
+/*
+ * SSL_get_error {{{2
+ * -------------
+ */
+int ossl_quic_get_error(const QUIC_CONNECTION *qc, int i)
+{
+    return SSL_ERROR_NONE; // TODO
+}
+
+/*
+ * SSL_write {{{2
+ * ---------
+ */
+struct quic_write_again_args {
+    QUIC_CONNECTION *qc;
+    const void      *buf;
+    size_t          len;
+    size_t          *written;
+};
+
+static int quic_write_again(void *arg)
+{
+    struct quic_write_again_args *args = arg;
+
+    // TODO what if connection is torn down due to error while blocking (e.g. rx
+    // bad frame)
+
+    if (!ossl_quic_sstream_append(args->qc->stream0->sstream, args->buf,
+                                   args->len, args->written))
+        return -2;
+
+    if (*args->written > 0)
+        /* appended at least one byte, the SSL_write op can finish now */
+        return 1;
+
+    return 0; /* did not write anything, keep trying */
 }
 
 int ossl_quic_write(SSL *s, const void *buf, size_t len, size_t *written)
 {
-    BIO *wbio = SSL_get_wbio(s);
-    int ret;
-    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_QUIC_SSL(s);
+    QUIC_CONNECTION *qc = QUIC_CONNECTION_FROM_SSL(s);
+    struct quic_write_again_args args;
+    int res;
 
-    if (sc == NULL || wbio == NULL)
+    *written = 0;
+
+    if (!EXPECT_QUIC_CONN(qc))
         return 0;
 
-    sc->rwstate = SSL_WRITING;
-    ret = BIO_write_ex(wbio, buf, len, written);
-    if (ret > 0 || !BIO_should_retry(wbio))
-        sc->rwstate = SSL_NOTHING;
-    return ret;
+    // TODO if closed
+    // TODO stop_sending, reset_stream?
+
+    if (qc->stream0 == NULL || qc->stream0->sstream == NULL)
+        return 0;
+
+    if (!ossl_quic_sstream_append(qc->stream0->sstream, buf,
+                                  len, written)) {
+        /* Stream already finished or allocation error. */
+        *written = 0;
+        return 0;
+    }
+
+    if (*written > 0) {
+        /*
+         * We have appended at least one byte to the stream.
+         * Potentially mark stream as active, depending on FC.
+         */
+        ossl_quic_stream_map_update_state(&qc->qsm, qc->stream0);
+
+        /*
+         * Try and send.
+         *
+         * TODO(QUIC): It is probably inefficient to try and do this
+         * immediately, plus we should eventually consider Nagle's algorithm
+         * etc.
+         */
+        reactor_tick(&qc->rtor);
+    } else if (blocking_mode(qc)) {
+        /*
+         * We were not able to append anything immediately, so our stream buffer
+         * is probably full. This means we need to block until some of it is
+         * freed up.
+         */
+        args.qc         = qc;
+        args.buf        = buf;
+        args.len        = len;
+        args.written    = written;
+
+        res = reactor_block_until_pred(&qc->rtor, quic_write_again, &args, 0);
+        if (res <= 0)
+            return 0;
```

Comment:
Shouldn't we loop and try and append the data again?

Reply:
This is what `reactor_block_until_pred` and `quic_write_again` does?